### ISSUE TITLE
models.py

### ISSUE THREAD

**Issue by karthik78437** [karthik78437/karthik#4]
"""
models.py

Defines baseline LSTM model and Attention-LSTM model using Keras.
The models are sequence-to-one regressors.
"""

from tensorflow.keras.models import Model
from tensorflow.keras.layers import Input, LSTM, Dense, Dropout, Bidirectional
from attention_layer import TemporalAttention

def build_baseline_lstm(input_shape,
                       lstm_units=64,
                       dropout=0.2,
                       bidirectional=False):
    """
    Build a baseline LSTM for sequence-to-one forecasting.

    Args:
        input_shape: (seq_len, n_features)
    Returns:
        model: compiled Keras model
    """
    seq_len, n_features = input_shape
    inp = Input(shape=(seq_len, n_features), name='input')
    if bidirectional:
        x = Bidirectional(LSTM(lstm_units, return_sequences=False))(inp)
    else:
        x = LSTM(lstm_units, return_sequences=False)(inp)
    x = Dropout(dropout)(x)
    out = Dense(1, activation='linear', name='output')(x)
    model = Model(inputs=inp, outputs=out, name='baseline_lstm')
    model.compile(optimizer='adam', loss='mse', metrics=['mae'])
    return model

def build_attention_lstm(input_shape,
                         lstm_units=64,
                         dropout=0.2,
                         attention_return_attention=False,
                         bidirectional=False):
    """
    Build LSTM with TemporalAttention.

    Args:
        input_shape: (seq_len, n_features)
    Returns:
        model: compiled Keras model that outputs prediction (and optionally attention).
    """
    seq_len, n_features = input_shape
    inp = Input(shape=(seq_len, n_features), name='input')
    # return sequences to feed into attention
    if bidirectional:
        x = Bidirectional(LSTM(lstm_units, return_sequences=True))(inp)
    else:
        x = LSTM(lstm_units, return_sequences=True)(inp)
    x = Dropout(dropout)(x)
    # Attention layer — returns context vector and optionally weights
    att_layer = TemporalAttention(return_attention=attention_return_attention)
    if attention_return_attention:
        context, att_w = att_layer(x)
        out = Dense(1, activation='linear', name='output')(context)
        model = Model(inputs=inp, outputs=[out, att_w], name='attention_lstm')
        # compile with custom loss mapping: primary loss for output
        model.compile(optimizer='adam', loss={'output': 'mse'}, metrics={'output': 'mae'})
    else:
        context = att_layer(x)
        out = Dense(1, activation='linear', name='output')(context)
        model = Model(inputs=inp, outputs=out, name='attention_lstm')
        model.compile(optimizer='adam', loss='mse', metrics=['mae'])
    return model
